---
output: github_document
---
<!-- README.md is generated from README.Rmd. Please edit that file -->

# `iProFun`

## An *i*ntegrative analysis tool to screen for *Pro*teogenomic *Fun*ctional traits perturbed by DNA copy number alterations (CNA) and DNA methylation


The goal of **iProFun** is to 

* characterize functional consequences of DNA copy number and methylation alterations in tumors 

* facilitate screening for cancer drivers contributing to tumor initiation and progression, since CNAs and DNA methylations that preserve functional consequences are more likely to be cancer drivers. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
load("Rdata/rna_normal.Rdata")
load("Rdata/phospho_normal.Rdata")
load("Rdata/protein_normal.Rdata")
load("Rdata/methy.Rdata")
load("Rdata/cnv.Rdata")
load("Rdata/rna_pc_1_3.Rdata")
load("Rdata/protein_pc_1_3.Rdata")
load("Rdata/phospho_pc_1_3.Rdata")
load("Rdata/cnv_input_1_3.RData")
load("Rdata/methy_input_1_3.RData")
```




### Installation

You can install:

<!-- * the most recent officially-released version from CRAN with -->

<!--     ```R -->
<!--     install.packages("iProFun") -->
<!--     ```` -->

* the latest development version from GitHub with

    ```R
    install.packages("devtools")
    devtools::install_github("xiaoyu/iProFun")
    ````
    
    
## iProFun Integrative analysis pipeline

Below is an example of how iProFun is commonly used.  A full description of the tool can be found in our MCP paper.

```{r}
library(iProFun)
require(metRology)
require(matrixStats)
```


### Data summary

After preprossing and data cleaning, we have 15121 genes and 569 subjects for mRNA data, 7010 genes and 174 subjects for protein, 5685 genes and 70 subjects for phospho data, 25762 genes and 552 subjects for methylation data, 11859 genes and 560 subjects for mRNA data. The following shows the data structure for each data.

```{r}
rna_normal[1:5,1:5]
protein_normal[1:5,1:5]
phospho_normal[1:5, 1:5]
methy[1:5,1:5]
cnv[1:5, 1:5]
```

### Gene-level multiple linear regression to obtain summary statistics

```{r, eval=FALSE}
ylist_normal = list(rna_normal, protein_normal, phospho_normal)
methy_input_1_3 <-
MultiReg_together(
ylist = ylist_normal,
xlist = list(methy, cnv),
covariates = list(rna_pc_1_3, protein_pc_1_3, phospho_pc_1_3),
cl = cl
)
cnv_input_1_3 <-
MultiReg_together(
ylist = ylist_normal,
xlist = list(cnv, methy),
covariates = list(rna_pc_1_3, protein_pc_1_3, phospho_pc_1_3),
cl = cl
)

```





```{r}
str(cnv_input_1_3)
str(methy_input_1_3)
```


### Primo â€“ An integrative analysis method for detecting joint associations of DNA al- terations with multi-omics traits

```{r, results=FALSE}
pi1=rep(0.05, 3)
cnv_1_3 = MultiOmics_Input(cnv_input_1_3, pi1=pi1)
methy_1_3 = MultiOmics_Input(methy_input_1_3, pi1=pi1)
```

```{r}
str(cnv_1_3)
str(methy_1_3)
```


### False discovery rate assessment
